### PR TITLE
CI: Enable ASAN testing

### PR DESCRIPTION
**Pull request by ivankochin**
## What
Add Address Sanitizer check job to CI

### CODE REVIEW

File: test/gtest/Makefile.am [hunk 1]
```diff
@@ -27,6 +28,12 @@ export UCX_HANDLE_ERRORS
 export UCX_LOG_LEVEL
 export UCX_LOG_PRINT_ENABLE
 
+if HAVE_ASAN
+	ENV = \
+		LSAN_OPTIONS=suppressions=$(top_srcdir)/contrib/lsan.supp \
+		ASAN_OPTIONS=protect_shadow_gap=0
```
Comment (yosefe) [inline test/gtest/Makefile.am:34]: why not set it as other env vars, for example UCX_LOG_PRINT_ENABLE in line 18

File: contrib/test_jenkins.sh [hunk 2]
```diff
@@ -1135,7 +1143,7 @@ then
     check_machine
     if [[ "$PROTO_ENABLE" == "no" ]]; then
         run_test_proto_disable
-    else
-        run_tests
+    elif [[ "$ASAN_CHECK" == "yes" ]]; then
+        run_asan_check
```
Comment (yosefe) [inline contrib/test_jenkins.sh:1147]: what about run_tests ?

File: config/m4/compiler.m4 [hunk 3]
```diff
@@ -424,8 +424,8 @@ AS_IF([test "x$enable_asan" = xyes],
                                      [-fsanitize=address -fno-omit-frame-pointer],
                                      [AC_LANG_SOURCE([[int main(int argc, char** argv){return 0;}]])],
                                      [AS_MESSAGE([compiling with sanitizer])
-                                      BASE_CXXFLAGS="-fsanitize=address -fno-omit-frame-pointer $BASE_CXXFLAGS"
-                                      LDFLAGS="-fsanitize=address -fno-omit-frame-pointer $LDFLAGS"],
+                                      BASE_CXXFLAGS="-fsanitize=address -static-libasan -fno-omit-frame-pointer $BASE_CXXFLAGS"
+                                      LDFLAGS="-fsanitize=address -static-libasan -fno-omit-frame-pointer $LDFLAGS"],
```
Comment (yosefe) [inline config/m4/compiler.m4:428]: why need static?

File: buildlib/pr/main.yml [hunk 4]
```diff
@@ -398,6 +398,19 @@ stages:
...
+    dependsOn: [Static_check]
+    jobs:
+    - template: tests.yml
+      parameters:
+        name: gpu
+        demands: ucx_gpu -equals yes
+        test_perf: 0
+        container: ubuntu22_cuda_12_1
+        asan_check: yes
+        valgrind_disable: yes
```
Comment (yosefe) [inline buildlib/pr/main.yml:412]: maybe change valgrind_disable to valigrind_check to make it consistent

File: test/gtest/Makefile.am (hunk 1 above)
Comment (ivankochin) [reply test/gtest/Makefile.am]: Done.

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (ivankochin) [reply contrib/test_jenkins.sh:1147]: Done.

File: buildlib/pr/main.yml (hunk 4 above)
Comment (ivankochin) [reply buildlib/pr/main.yml:412]: If you don't object, I would prefer to make it in a separate PR.

Comment (yosefe) [reply buildlib/pr/main.yml:412]: ok

File: contrib/configure-devel [hunk 5]
```diff
@@ -14,7 +14,6 @@ basedir=$(cd $(dirname $0) && pwd)
 $basedir/../configure \
 	--enable-gtest \
 	--enable-examples \
-	--with-valgrind \
```
Comment (yosefe) [inline contrib/configure-devel:17]: pls keep it. gtest can do --without-valgrind

File: config/m4/compiler.m4 [hunk 6]
```diff
@@ -435,6 +435,7 @@ AS_IF([test "x$enable_asan" = xyes],
                     )],
       [])
 
+AM_CONDITIONAL([HAVE_ASAN], [test "x$enable_asan" != xno])
```
Comment (yosefe) [inline config/m4/compiler.m4:438]: remove

File: configure.ac [hunk 7]
```diff
@@ -179,6 +179,7 @@ AS_IF([test "x$with_docs_only" = xyes],
      UCX_DX_ENABLE_CHECK([man])
      UCX_DX_ENABLE_CHECK([pdf])
      AM_CONDITIONAL([DOCS_ONLY], [true])
+     AM_CONDITIONAL([HAVE_ASAN], [false])
```
Comment (yosefe) [inline configure.ac:182]: remove

File: test/gtest/Makefile.am [hunk 8]
```diff
@@ -27,6 +28,11 @@ export UCX_HANDLE_ERRORS
 export UCX_LOG_LEVEL
 export UCX_LOG_PRINT_ENABLE
 
+if HAVE_ASAN
```
[the remark this answers is not in the document]
Comment (yosefe) [reply test/gtest/Makefile.am:31]: remove

File: test/gtest/Makefile.am [hunk 9]
```diff
@@ -18,6 +18,7 @@ UCX_LOG_LEVEL            ?= warn
 UCX_LOG_PRINT_ENABLE     ?= y
 GTEST_FILTER             ?= *
 GTEST_EXTRA_ARGS         ?=
+ENV                      ?=
```
[the remark this answers is not in the document]
Comment (yosefe) [reply test/gtest/Makefile.am:21]: remove

File: contrib/test_jenkins.sh [hunk 10]
```diff
@@ -1070,6 +1067,13 @@ run_tests() {
 	export UCX_IB_ROCE_LOCAL_SUBNET=y
 	export UCX_IB_ROCE_SUBNET_PREFIX_LEN=inf
 
+}
+
+#
+# Run all tests
+#
+run_tests() {
+	set_ucx_common_test_env
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply contrib/test_jenkins.sh]: Done.

File: test/gtest/Makefile.am (hunk 9 above)
[the remark this answers is not in the document]
Comment (ivankochin) [reply test/gtest/Makefile.am]: Done.

File: test/gtest/Makefile.am (hunk 8 above)
[the remark this answers is not in the document]
Comment (ivankochin) [reply test/gtest/Makefile.am]: Done.

File: contrib/configure-devel (hunk 5 above)
Comment (ivankochin) [reply contrib/configure-devel]: Done.

File: config/m4/compiler.m4 (hunk 6 above)
Comment (ivankochin) [reply config/m4/compiler.m4]: Done.

File: configure.ac (hunk 7 above)
Comment (ivankochin) [reply configure.ac]: Done.

**Review by yosefe — APPROVED** [review]

**Review by iyastreb — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

File: buildlib/pr/main.yml [hunk 11]
```diff
@@ -168,6 +168,9 @@ resources:
     - container: ubuntu2004_rocm_5_4_0
       image: rdmz-harbor.rdmz.labs.mlnx/ucx/x86_64/ubuntu2004:rocm_5_4_0
       options: $(DOCKER_OPT_ARGS) $(DOCKER_OPT_VOLUMES)
+    - container: ubuntu22_cuda12
+      image: rdmz-harbor.rdmz.labs.mlnx/ucx/x86_64/ubuntu22.04-mofed5-cuda12:3
+      options: $(DOCKER_OPT_ARGS) $(DOCKER_OPT_VOLUMES) $(DOCKER_OPT_GPU)
```
Comment (ivankochin) [inline buildlib/pr/main.yml:173]:
```suggestion
      options: $(DOCKER_OPT_ARGS) --net=host $(DOCKER_OPT_VOLUMES) $(DOCKER_OPT_GPU)
```

File: buildlib/pr/main.yml [hunk 12]
```diff
@@ -170,7 +170,7 @@ resources:
       options: $(DOCKER_OPT_ARGS) $(DOCKER_OPT_VOLUMES)
     - container: ubuntu22_cuda12
       image: rdmz-harbor.rdmz.labs.mlnx/ucx/x86_64/ubuntu22.04-mofed5-cuda12:3
-      options: $(DOCKER_OPT_ARGS) $(DOCKER_OPT_VOLUMES) $(DOCKER_OPT_GPU)
+      options: $(DOCKER_OPT_ARGS) --net=host $(DOCKER_OPT_VOLUMES) $(DOCKER_OPT_GPU)
```
[the remark this answers is not in the document]
Comment (Alexey-Rivkin) [reply buildlib/pr/main.yml:173]: We have 2 Docker containers with the "host" network mode:
 `centos7_ib`
 `centos7_cuda11`
https://github.com/openucx/ucx/blob/master/buildlib/pr/main.yml#L12-L17
Others default to the "bridge" network mode.

[the remark this answers is not in the document]
Comment (yosefe) [reply buildlib/pr/main.yml:173]: @Alexey-Rivkin thanks. Basically, all dockers that run gtest should be with "--net host". I guess most dockers are build only.

**Review by yosefe — APPROVED** [review]

File: contrib/test_jenkins.sh [hunk 13]
```diff
@@ -1120,14 +1116,26 @@ run_test_proto_disable() {
 	run_gtest "default"
 }
 
+run_asan_check() {
+	# build for devel tests and gtest
+	build devel --enable-gtest --enable-asan --without-valgrind
+
+	# all are running gtest
```
Comment (brminich) [inline contrib/test_jenkins.sh:1123]: minor: comments are not really informative here (copy-past)? I'd remove

File: test/gtest/common/test_helpers.cc [hunk 14]
```diff
@@ -326,6 +326,9 @@ int test_time_multiplier()
     int factor = 1;
 #if _BullseyeCoverage
     factor *= 10;
+#endif
```
Comment (brminich) [inline test/gtest/common/test_helpers.cc:329]: maybe
```
   int factor = 1;
    if (RUNNING_ON_VALGRIND) {
        factor *= 20;
    }
#if _BullseyeCoverage
    factor *= 10;
#endif
#ifdef __SANITIZE_ADDRESS__
    factor *= 20;
#endif

    return factor;
```

Comment (ivankochin) [reply test/gtest/common/test_helpers.cc:332]: Done.

File: contrib/test_jenkins.sh (hunk 13 above)
Comment (ivankochin) [reply contrib/test_jenkins.sh]: Done.

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**